src/psbt/finalizer.rs

```diff
@@ -0,0 +1,371 @@
+            });
+        }
+        if redeem_script.is_v0_p2wsh() {
+            // 5. `ShWsh` case
+            if let Some(ref witness_script) = inp.witness_script {
+                if witness_script.to_v0_p2wsh() != *redeem_script {
+                    return Err(InputError::InvalidWitnessScript {
+                        witness_script: witness_script.clone(),
+                        p2wsh_expected: redeem_script.clone(),
+                    });
+                }
+                let ms = Miniscript::<bitcoin::PublicKey, Segwitv0>::parse(witness_script)?;
+                return Ok(Descriptor::ShWsh(ms));
+            } else {
+                return Err(InputError::MissingWitnessScript);
+            }
+        } else if redeem_script.is_v0_p2wpkh() {
+            // 6. `ShWpkh` case
+            let partial_sig_contains_pk = inp
+                .partial_sigs
+                .iter()
+                .filter(|&(&pk, _sig)| {
+                    let addr = bitcoin::Address::p2wpkh(&pk, bitcoin::Network::Bitcoin)
+                        .expect("Address corresponding to valid pubkey");
+                    *script_pubkey == addr.script_pubkey()
+                })
+                .next();
+            match partial_sig_contains_pk {
+                Some((pk, _sig)) => {
+                    return Ok(Descriptor::ShWpkh(pk.to_owned()));
+                }
+                None => return Err(InputError::MissingPubkey),
+            }
+        } else {
+            //7. regular p2sh
+            if inp.witness_script.is_some() {
+                return Err(InputError::NonEmptyWitnessScript);
+            }
+            if let Some(ref redeem_script) = inp.redeem_script {
+                let ms = Miniscript::<bitcoin::PublicKey, Legacy>::parse(redeem_script)?;
+                return Ok(Descriptor::Sh(ms));
+            } else {
+                return Err(InputError::MissingWitnessScript);
+            }
+        }
+    } else {
+        // 8. Bare case
+        if inp.witness_script.is_some() {
+            return Err(InputError::NonEmptyWitnessScript);
+        }
+        if inp.redeem_script.is_some() {
+            return Err(InputError::NonEmptyRedeemScript);
+        }
+        let ms = Miniscript::<bitcoin::PublicKey, Legacy>::parse(script_pubkey)?;
+        return Ok(Descriptor::Bare(ms));
+    }
+}
+
+/// Interprets all psbt inputs and checks whether the
+/// script is correctly interpreted according to the context
+/// The psbt must have included final script sig and final witness.
+/// In other words, this checks whether the finalized psbt interprets
+/// correctly
+pub fn interpreter_check(psbt: &Psbt) -> Result<(), Error> {
+    for (index, input) in psbt.inputs.iter().enumerate() {
+        let script_pubkey =
+            get_scriptpubkey(psbt, index).map_err(|e| Error::InputError(e, index))?;
+        let empty_script_sig = Script::new();
+        let empty_witness = Vec::new();
```

Comment:
nit: i would drop both of these variables, they are only used once and both have longer names than the expressions they replace :P

Reply:
I think I need a let binding for the following?

`fn_name(&Script::new())`